### PR TITLE
Add multi-thread inference example which shares the inference_program and parameters

### LINKED ISSUE #9650: Support multi-thread inference for the same program and sharing the parameters
Sometimes users want to use multi-thread to do inference for the same program, so they want to share the parameters among different threads to minimize the memory usage.

In Fluid, it is easy to share variables among threads, the only need is to run with the same `scope` and the same type of `place`.

Both parameters and feed, fetch holder variables are defined as `persistable`. However, we do not want to share feed, fetch holder variables among different threads, or the inference results will be overridden by the slowest thread's computing result. If the program defined in main threads contains `feed_op`s and `fetch_op`s, we need clone a copy of the program and change the `feed/fetch_holder_name` to a unique one in each thread.

### PR DESCRIPTION
**Pull request by Xreki**
Fix #9650

### CODE REVIEW

File: paddle/fluid/framework/program_desc.cc [hunk 1]
```diff
@@ -96,15 +96,53 @@ const std::vector<std::string> ProgramDesc::GetFeedTargetNames() {
...
-  for (auto *op : global_block->AllOps()) {
+  for (auto *op : global_block.AllOps()) {
     if (op->Type() == kFetchOpType) {
       fetch_target_names.push_back(op->Input("X")[0]);
     }
   }
   return fetch_target_names;
 }
 
+void ProgramDesc::SetFeedHolderName(const std::string &feed_holder_name) {
```
Comment (Superjomn) [inline paddle/fluid/framework/program_desc.cc]: Is feed/fetch holder a new concept? add some comments will be better.

File: paddle/fluid/framework/program_desc.h [hunk 2]
```diff
@@ -55,6 +56,9 @@ class ProgramDesc {
   const std::vector<std::string> GetFeedTargetNames();
   const std::vector<std::string> GetFetchTargetNames();
 
+  void SetFeedHolderName(const std::string &feed_holder_name);
```
Comment (Superjomn) [inline paddle/fluid/framework/program_desc.h]: Is feed/fetch holder a new concept? add some comments will be better.

Comment (Xreki) [reply paddle/fluid/framework/program_desc.h]: Done.

File: paddle/fluid/inference/tests/test_multi_thread_helper.h [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    fetch_targets[fetch_target_names[i]] = cpu_fetchs[i];
+  }
+
+  // 6. Run the inference program
+  executor->Run(*copy_program, scope, feed_targets, fetch_targets,
+                feed_holder_name, fetch_holder_name);
+}
+
+template <typename Place>
+void TestMultiThreadInference(
```
[the remark this answers is not in the document]
Comment (Xreki) [reply paddle/fluid/inference/tests/test_multi_thread_helper.h]: 多线程作为一种特殊的应用场景，放在一个单独的文件里面，方便索引和引用吧。

File: paddle/fluid/inference/tests/book/test_inference_fit_a_line.cc [hunk 4]
```diff
@@ -49,9 +51,72 @@ TEST(inference, fit_a_line) {
...
   // Run inference on CUDA GPU
+  LOG(INFO) << "--- GPU Runs: ---";
   TestInference<paddle::platform::CUDAPlace>(dirname, cpu_feeds, cpu_fetchs2);
   LOG(INFO) << output2.dims();
 
   CheckError<float>(output1, output2);
 #endif
 }
+
+TEST(multi_thread_inference, fit_a_line) {
```
[the remark this answers is not in the document]
Comment (Xreki) [reply paddle/fluid/inference/tests/book/test_inference_fit_a_line.cc]: Done.

**Review by luotao1 — APPROVED** [review]
LGTM!

**Pull request merged**